docs/design-documents/20220309-opencdc.md

```diff
@@ -0,0 +1,349 @@
+# OpenCDC
+
+## Introduction
+
+### Goals
+
+The main goal is to ensure compatibility between any combination of source/destination connectors.
+
+Define a format that fulfills the following criteria:
+- A record can be associated to a schema that describes the payload or key
+- The paylod/key schema can be included in the record or pointing to a schema stored in a schema registry
+- Be able to represent an inserted, updated and deleted record
+- Be able to represent a record created in a snapshot
+- Support for splitting big records into multiple parts
+- Define standard metadata fields
+
+Besides the format itself, we want to roughly define how Conduit will work with the record, specifically:
+- How transforms change the associated schema?
+- How can a schema be attached to a raw record?
+- How can a schema be extracted to a schema registry?
+
+### Questions
+
+- **Should we have a separate field announcing the record was created as part of a snapshot?**
+  - No, this information will be stored in the action (insert, update, delete, snapshot).
+- **Should the record support sending a schema only (e.g. we transmit a CREATE TABLE statement)? This might be useful for setting up the target structure even if the source is still empty.**
+  - No, for now we ignore this case.
```

Comment:
After some more thought, I think it boils down to how we want to send schema changes over the wire. Would they get their own message followed by records that adhere to that schema update, or would schema changes be sent along with the record that adheres to them and the reader would have to detect and apply the schema changes?

Reply:
Emitting schemas in the form of DDL statements (e.g. a `CREATE TABLE` statement) only makes sense in the case where you're doing straight replication between identical resources. 

As soon as you mix connectors and/or introduce transforms this functionality becomes useless.

I think what we actually want is the ability to detect a schema from a given Source and _register_ it with a schema registry somewhere. This would then allow the destination to preemptively query that registry and create the table structure in advance, however this also falls apart if a transform is introduced.

That is to say, all of the above leads me to think that actually the lazy schema lookup/integration is probably the most sensible approach. E.g. when a destination sees a record with a schema (ID or version) that differs from what it currently has set, then it can _infer_ the DDL to support the new schema.

Reply:
> when a destination sees a record with a schema (ID or version) that differs from what it currently has set, then it can infer the DDL to support the new schema.

The record design includes `before` and `after` data, possibly even schemas for both, so this should be doable. We may want to provide utilities for detecting schema changes in the SDK to make it easier for the connector developer.

Reply:
Since keys and payloads in both `before` and `after` have schemas, we have a complete picture of a record's change at any point in time. This leads to the question of how diffs should be represented and handled but I think that's out of the scope of this document.

I do agree that the lazy lookup / registration approach maximizes flexibility, and I think that's what we want for Conduit's many-to-many ecosystem.

>  Emitting schemas in the form of DDL statements (e.g. a CREATE TABLE statement) only makes sense in the case where you're doing straight replication between identical resources. 

Additionally, adding command language or anything similar in records will spiral out of control in complexity. I would say we should avoid any approach that would have us sending DDL statements over the wire.